src/plugins/fs_routes/mod.ts

```diff
@@ -62,60 +64,67 @@ export interface FsRoutesOptions {
   loadIsland: (path: string) => Promise<unknown>;
 }
 
-export interface TESTING_ONLY__FsRoutesOptions {
-  _fs?: FsAdapter;
-}
-
 export async function fsRoutes<State>(
   app: App<State>,
-  options_: FsRoutesOptions,
+  options: FsRoutesOptions,
 ) {
-  const options = options_ as FsRoutesOptions & TESTING_ONLY__FsRoutesOptions;
-  const ignore = options.ignoreFilePattern ?? [TEST_FILE_PATTERN];
-  const fs = options._fs ?? fsAdapter;
+  const skip = options.ignoreFilePattern ?? [TEST_FILE_PATTERN];
 
   const dir = options.dir
-    ? parseRootPath(options.dir, fs.cwd())
+    ? parseRootPath(options.dir, Deno.cwd())
     : app.config.root;
   const islandDir = path.join(dir, "islands");
   const routesDir = path.join(dir, "routes");
 
   const islandPaths: string[] = [];
   const relRoutePaths: string[] = [];
 
-  // Walk routes folder
-  await Promise.all([
-    walkDir(
-      islandDir,
-      (entry) => {
-        islandPaths.push(entry.path);
-      },
-      ignore,
-      fs,
-    ),
-    walkDir(
-      routesDir,
-      (entry) => {
-        const relative = path.relative(routesDir, entry.path);
-
-        // A `(_islands)` path segment is a local island folder.
-        // Any route path segment wrapped in `(_...)` is ignored
-        // during route collection.
-        const match = relative.match(GROUP_REG);
-        if (match && match[2][0] === "_") {
-          if (match[2] === "_islands") {
-            islandPaths.push(entry.path);
-          }
-          return;
+  try {
+    for await (
+      const entry of walk(islandDir, {
+        includeDirs: false,
+        exts: ["tsx", "jsx", "ts", "js"],
+        skip,
+      })
+    ) {
+      islandPaths.push(entry.path);
+    }
+  } catch (error) {
+    if (!(error instanceof Deno.errors.NotFound)) {
+      throw error;
+    }
+  }
+
+  try {
+    for await (
+      const entry of internals.walk(routesDir, {
+        includeDirs: false,
+        exts: ["tsx", "jsx", "ts", "js"],
+        skip,
+      })
+    ) {
+      const relative = path.relative(routesDir, entry.path);
+
+      // A `(_islands)` path segment is a local island folder.
+      // Any route path segment wrapped in `(_...)` is ignored
+      // during route collection.
+      const match = relative.match(GROUP_REG);
+      if (match && match[2][0] === "_") {
+        if (match[2] === "_islands") {
+          islandPaths.push(entry.path);
         }
+        continue;
+      }
 
-        const url = new URL(relative, "http://localhost/");
-        relRoutePaths.push(url.pathname.slice(1));
-      },
-      ignore,
-      fs,
-    ),
-  ]);
+      const url = new URL(relative, "http://localhost/");
+      relRoutePaths.push(url.pathname.slice(1));
+    }
+  } catch (error) {
+    // `islandDir` or `routesDir` does not exist, so we can skip it
+    if (!(error instanceof Deno.errors.NotFound)) {
+      throw error;
+    }
+  }
```

Comment:
Here, instead of checking whether the first argument passed to `walk()` is a directory, we just simply try to walk it. If it's not a directory, a `Deno.errors.NotFound` error is thrown and we ignore it. This has the same functionality as before, but just without having to use some `isDirectory()` function.

Reply:
Is there any performance gain from avoiding the isDirectory function?

Reply:
Perhaps a slight performance benefit. But that's not the goal of the change. It just avoids the potential for hitting a (very unlikely) race condition by just attempting the operation and catching if it fails instead of checking if the operation can be performed then doing the operation if it can be. Does that make sense?

This was the exact topic of discussion for `fs.exists()` within the Standard Library. See the note in https://jsr.io/@std/fs/doc/~/exists.